### ISSUE TITLE
Improve IACT DL3 data access

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1303]
Currently DL3 data in Gammapy is always accessed via the index tables and the [gammapy.data.DataStore](http://docs.gammapy.org/dev/api/gammapy.data.DataStore.html), as shown in [this tutorial](http://docs.gammapy.org/dev/notebooks/cta_1dc_introduction.html#Index-files-and-DataStore).

The interface to analysis code in Gammapy is currently either

* A [gammapy.data.ObservationList](http://docs.gammapy.org/dev/api/gammapy.data.ObservationList.html) containing [gammapy.data.DataStoreObservation](http://docs.gammapy.org/dev/api/gammapy.data.DataStoreObservation.html) objects, that have an `obs_id` and reference back to a [gammapy.data.DataStore](http://docs.gammapy.org/dev/api/gammapy.data.DataStore.html) or
* A list (or equivalently 1D numpy array) of `obs_id` and `DataStore`

---

@registerrier @mackaiver @adonath and I (and probably others) have discussed a few times that we should improve this, but I couldn't find an issue. So let's discuss here for a bit, and then hopefully someone follows up with code improvements.

Note that #1265 is already open and is just a proposal to remove the "master index" and "DataManager" from Gammapy for now. This is even higher level than what is discussed here, and can be discussed / decided independently.

---

I think all of our previous discussions concluded that we need something similar to our existing `ObservationList` and `DataStoreObservation` class, but that they should be re-written in a more flexible way that in addition to what it does now for us:
* be created in a simpler way, without having to generate index tables
* independent of DataStore - the data store and index tables remain and are very useful if you have 1000s of runs, but they are just one way to create IACT observation and observations objects.
* ideally we will support other types of observations also, e.g. Fermi-LAT, IACT slewing, ..., i.e. the observations interface should be generic and simple.

This is similar to what Gammalib / ctools has:
http://gammalib.sourceforge.net/user_manual/modules/obs.html

Concretely I think we should:
* put new, better code in gammapy.data in place, with new names like e.g. `Observations` and `Observation`. These would be ~ 100 lines of code, and later would probably grow to a few 100 as we support more types of observations and some support conveniences for analysis (e.g. object mappers to pointing position as `SkyCoord` and store such properties on the observation object).
* deprecate the existing `ObservationList` and `DataStoreObservation` and switch the `DataStore` over to use the new classes

Since almost no-one outside Gammapy itself is importing `ObservationList` and `DataStoreObservation` directly, another option is to evolve these two classes in-place to the new, better solution. But I think the "replace by new better code" here would work more nicely.

**Comment by joleroi** [comment]
I talked to @dcfidalgo about the status of ``gammapy.data`` today. Later I found this issue describing the most important points already. I think what was suggested above is in scope for the gammapy coding sprint. 

The first step would be introducing an ``Observation`` class that takes an event list and IRFs on init and is completely independent of a ``DataStore``. This class will be exposed to users who want to develop/prototype new methods and can also be the basis for creating new observations e.g. for lightcurve or pulsar analysis.

Later the ``DataStore`` need to be connected to this class to allow a memory efficient analysis of O(100) runs. I would keep the ``DataStoreObservation`` but let it be a daughter class of ``Observation`` with all attributes set to ``None`` on init and add the functionality to load the data from disk when a certain attribute is accessed.

____

And additional point which would be nice to have is the possibility to create a ``DataStore`` in a programmatic way. After generating a list of ``Observations`` in a script it should be easily possible to write those observations to disk and create a default version of the index files. This would be very e.g. for a lightcurve analysis of many runs. The data could be loaded in chunks from an official FITS release, modified to contain the correct GTIs and written back to disk for the actual analysis. The same functionality would also allow users to take an official release but simulate some part of it (event list, IRFs).

**Comment by cdeil** [comment]
@joleroi @dcfidalgo - Awesome!

> I think what was suggested above is in scope for the gammapy coding sprint.

Yes and no. If you're adding a new class or options, then yes. But touching many callers and lines during that week is not a good idea, because there will be ~ 15 people hacking on Gammapy that week, so everyone must make small and quick and focused PRs that don't tough too many lines / files, or it'll be chaos. So try to find a series of small PRs that iterate towards a better future, and consider making some changed the week before or only after the coding sprint.

**Comment by joleroi** [comment]
> If you're adding a new class or options, then yes.

That's the plan

**Comment by cdeil** [comment]
Closing this old issue. I think it's OK now!?
